src/main/java/com/mathworks/ci/utilities/MatlabCommandRunner.java

```diff
@@ -0,0 +1,215 @@
+                + File.separator 
+                + ".matlab");
+        matlabFolder.mkdirs();
+
+        // Create temp folder
+        this.tempFolder = matlabFolder.createTempDir("tempDir", null);
+        
+        // If we hit an error during shutdown while cleaning up
+        // there's not too much that we can do.
+        Runtime.getRuntime().addShutdownHook(
+                new Thread(() -> {
+                    try {
+                        tempFolder.deleteRecursive();
+                    } catch(Exception e) {
+                        System.err.println(e.toString());
+                    }
+                }));
+    }
+
+    /** 
+     * Spawns a process to run the specified command.
+     *
+     * @param command The command to run
+     */
+    public void runMatlabCommand(String command) throws IOException, InterruptedException, MatlabExecutionException {
+
+        this.params.getTaskListener().getLogger()
+            .println("\n#################### Starting command output ####################");
+
+        // Prepare the executable
+        FilePath exePath = prepareRunnerExecutable();
+
+        // Create the script file
+        FilePath scriptFile = createFileWithContent(command);
+        String cmd = "setenv('MW_ORIG_WORKING_FOLDER', cd('"
+            + this.tempFolder.getRemote()
+            + "'));"
+            + scriptFile.getBaseName();
+
+        // Create command
+        ArgumentListBuilder args = new ArgumentListBuilder();
+        args.add(exePath.getRemote());
+        args.add(cmd);
+        args.add(this.params.getStartupOptions().split(" "));
+
+        // Add custom environment vars
+        EnvVars env = getEnvVars();
+        Utilities.addMatlabToEnvPathFrmAxis(
+                Computer.currentComputer(), 
+                this.params.getTaskListener(), 
+                env);
+
+        ProcStarter proc = this.params.getLauncher().launch()
+            .envs(env)
+            .cmds(args);
+        if (this.stdOut == null) {
+            proc.stdout(this.params.getTaskListener());
+        } else {
+            proc.stdout(this.stdOut);
+        }
+        proc.pwd(this.params.getWorkspace());
+
+        int code = proc.join();
+
+        if (code != 0) {
+            throw new MatlabExecutionException(code);
+        }
+    }
+
+    /**
+     * Redirects stdout.
+     *
+     * @param out the OutputStream to write to
+     */
+    public void redirectStdOut(OutputStream out) {
+        this.stdOut = out;
+    }
+
+    /**
+     * Adds an environment variable.
+     *
+     * @param key the environment variable name
+     * @param value the environment variable value
+     */
+    public void addEnvironmentVariable(String key, String value) {
+        additionalEnvVars.put(key, value);
+    }
+
+    public EnvVars getEnvVars() {
+        EnvVars env = new EnvVars(this.params.getEnvVars());
+        env.putAll(additionalEnvVars); 
+        return env;
+    }
```

Comment:
I sort of don't understand why we're doing this. It looks like we wrap environment variables twice: once in our parameters class definition and once in our runner class definition. Is this mostly for testing? It makes things a little confusing for me below when we create the script.

Reply:
Great question! The simple answer is that I hit sporadic concurrency issues in the pipeline steps when I tried to rely on one set of environment variables stored in the parameters. The issue was due to one piece of the code trying to add environment variables while another location was reading the environment variables field.

By having a separate mutable set of environment variables owned solely by the `MatlabCommandRunner` and then joining those with the "master" copy stored inside of the parameters object and returning it as a new object we can avoid any concurrent access violations.

For a little bit I was serializing the parameters object rather than constructing it within the `StepExecution` class, so it's likely that concurrent access is no longer an issue (I think it's likely the serialization process was what was reading the parameters object's fields while the build action was adding an environment variable), but I still like keeping the parameter objects totally immutable to make it easier for contributors in the future to avoid running into the same issues.